pennylane/transforms/control.py

```diff
@@ -0,0 +1,191 @@
+from pennylane.transforms.adjoint import adjoint
+
+
+def expand_with_control(tape, control_wire):
+    """Expand a tape to include a control wire on all queued operations.
+
+    Args:
+        tape: A QuantumTape.
+        control_wire: A single wire to use as the control wire.
+    Returns:
+        A new QuantumTape with the controlled operations.
+    """
+    with QuantumTape(do_queue=False) as new_tape:
+        for op in tape.operations:
+            if op.__class__ in CONTROL_MAPS:
+                # Create the controlled version of the operation
+                # and add that the to the tape context.
+                # NOTE(chase): Should we use type(op) here instead?
+                CONTROL_MAPS[op.__class__](op, control_wire)
+            else:
+                raise NotImplementedError(
+                    f"Control transform of operation {type(op)} is not registered."
+                    "You can define the control transform by using qml.register_control"
+                    "See documentation for more details."
+                )
+    return new_tape
+
+
+class ControlledOperation(Operation):
+    """A Controlled Operation.
+
+    Unless you are a Pennylane plugin developer, **you should NOT directly use this class**,
+    instead, use the ``qml.ctrl`` method.
+
+    The ``ControlledOperation`` is a container class that defines a set of operations that
+    should by applied relative to a single control wire or a list of control wires.
+
+    Certain simulators and quantum computers can take advantage of the controlled gate sparsity,
+    while other devices must rely on the op-by-op decomposition defined by the ``op.expand``
+    method.
+
+    Attributes:
+        subtape (QuantumTape): The tape that defines the underlying operation.
+        control_wires (Wires): The control wires.
+
+    Args:
+        tape: A QuantumTape. This tape defines the unitary that should be applied relative
+            to the control wires.
+        control_wires: A wire or set of wires.
+    """
+
+    par_domain = "A"
+    num_wires = AnyWires
+    num_params = property(lambda self: self.subtape.num_params)
+
+    def __init__(self, tape, control_wires, do_queue=True):
+        self.subtape = tape
+        self.control_wires = Wires(control_wires)
+        wires = self.control_wires + tape.wires
+        super().__init__(*tape.get_parameters(), wires=wires, do_queue=do_queue)
+
+    def expand(self):
+        tape = self.subtape
+        for wire in self.control_wires:
+            tape = expand_with_control(tape, wire)
+        # TODO(chase): Do we need to re-queue these ops?
+        return tape
+
+    def adjoint(self, do_queue=False):
+        def requeue_tape(tape):
+            for op in tape.operations:
+                op.queue()
+
+        with QuantumTape(do_queue=False) as new_tape:
+            # Execute all ops adjointed.
+            adjoint(requeue_tape)(self.subtape)

```

Comment:
I wonder if it's worth breaking `adjoint` into two functions - a tape transform, and a 'function' transform --- in cases like this, the tape transform could be used.

Similar to how the metric tensor exists as a tape transform _and_ a QNode transform, with the latter making use of the former.

Reply:
That is a good idea. There are some changes I want to do to the adjoint structure right now, so i think it would make sense to do it in that PR.